R/Spatial.R

```diff
@@ -0,0 +1,180 @@
+  Q3_Final_Counts <- Q3_Divided_Counts * compositions::geometricmean(Q3_Sample_Counts)
+  
+  #stash into seurat object under new assay
+  seuratObj[['Q3']] <- Seurat::CreateAssayObject(as(Q3_Final_Counts, "dgCMatrix"))
+  
+  return(seuratObj)
+}
+
+#' @title RUVg_Housekeeping_Normalization
+#'
+#' @description a normalization method that uses residuals from a fitted model to Remove Unwanted Variation by leveraging housekeeping genes (https://www.nature.com/articles/nbt.2931).
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @param k the number of factors of unwanted variation. See section "Choice of number of factors of unwanted variation." in paper cited in the description.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the RUVg assay. 
+#' @export
+RUVg_Housekeeping_Normalization <- function(seuratObj, assay = "RNA", k = 1){
+  housekeeping_genes <- c("ABCF1", "ACTB", "ALAS1", "B2M", "CLTC", "G6PD", "GAPDH", 
+                          "GUSB", "HPRT1", "LDHA", "PGK1", "POLR1B", "POLR2A",
+                          "RPL19", "RPLP0", "SDHA", "TBP", "TUBB")
+  counts <- as.matrix(seuratObj[[assay]]@counts)
+  rounded_counts <- round(as.matrix(counts))
+  RUVg <- RUVSeq::RUVg(rounded_counts,
+                       housekeeping_genes[housekeeping_genes %in% rownames(seuratObj)], 
+                       k = k)
+  seuratObj[['RUVg']] <- Seurat::CreateAssayObject(as(RUVg$normalizedCounts, "dgCMatrix"))
+  return(seuratObj)
+}
+
+#' @title NanoString_Housekeeping_Normalization
+#'
+#' @description a normalization method that uses the housekeeping genes to generate size factors for normalization.
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the "Housekeeping" assay. 
+#' @export
+
+NanoString_Housekeeping_Normalization <- function(seuratObj, assay = "RNA"){
+  
+  #Normalization Steps from Nanostring 
+  #1.  Calculate the geometric mean of the selected housekeeping genes for each lane.
+  #2. Calculate the arithmetic mean of these geometric means for all sample lanes.
+  #3. Divide this arithmetic mean by the geometric mean of each lane to generate a lane-specific normalization factor.
+  #4. Multiply the counts for every gene by its lane-specific normalization factor.
+  
+  counts <- as.matrix(seuratObj[[assay]]@counts) 
+  
+  housekeeping_genes <- c("ABCF1", "ACTB", "ALAS1", "B2M", "CLTC", "G6PD", "GAPDH", 
+                          "GUSB", "HPRT1", "LDHA", "PGK1", "POLR1B", "POLR2A",
+                          "RPL19", "RPLP0", "SDHA", "TBP", "TUBB")
+  housekeeping_counts <- counts[rownames(counts) %in% housekeeping_genes, ]
```

Comment:
should this die if the intersect is too low ? or at least report the # intersecting?

Reply:
agreed on both points. It should require at least one housekeeping gene and report how many are present.